### PR TITLE
Replace #[mantle::service] with compiler plugin

### PR DESCRIPTION
**Pull request by nhynes**
- [x] impl
- [x] remove unused code
- [x] update tests
    - [x] idl_gen
    - [x] xcc
    - [x] compiletests
- [x] new compiletests for idl-gen errors
- [~] update `mantle-test`
- [x] update examples
    - [x] ballot
    - [x] messaging

Still to-do:
* complete mantle-test
* allow defining `Service` structs outside of `main.rs`
* client generation

### CODE REVIEW

**Comment by nhynes** [comment]
@armaniferrante @willscott this is ready for preliminary review. I still have to port the tests over to use the compiler plugin.

File: mantle-build/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# mantle-build
+
+[![Latest Version](https://img.shields.io/crates/v/mantle-build.svg)](https://crates.io/crates/mantle-build)
+[![docs](https://docs.rs/mantle-build/badge.svg)](https://docs.rs/mantle-build)
+
+A compiler plugin that generates JSON descriptions of an RPC interface from Rust types.
```
Comment (willscott) [inline mantle-build/README.md]: can you expand on this a bit more since it does more than just JSON generation

File: README.md [hunk 2]
```diff
@@ -7,7 +7,7 @@ You can find more information in the [developer documentation](https://docs.oasi
 
 * **mantle** - A framework for conveniently developing blockchain services.
 * **mantle-test** - A testing framework for testing services built using Mantle.
-* **idl-gen** - A compiler plugin that generates RPC schemas from Rust types.
+* **mantle-build** - A compiler plugin that generates RPC schemas from Rust types.
```
Comment (nhynes) [inline README.md]:
```suggestion
* **mantle-build** - A compiler plugin that adds boilerplate code and generates RPC schemas from Rust types.
```

File: mantle-build/src/main.rs [hunk 3]
```diff
@@ -1,12 +1,12 @@
 //! Compiles a Mantle executable and generates the RPC interface definition.
-//! Usage: `GEN_IDL_FOR=<crate_name> IDL_TARGET_DIR=<dir> RUSTC_WRAPPER=idl-gen cargo build`
+//! Usage: `GEN_IDL_FOR=<crate_name> IDL_TARGET_DIR=<dir> RUSTC_WRAPPER=mantle-build cargo build`
```
Comment (nhynes) [inline mantle-build/src/main.rs]:
```suggestion
//! Usage: `GEN_IDL_FOR=<crate_name> RUSTC_WRAPPER=mantle-build cargo build`
```

File: examples/ballot/src/lib.rs [hunk 4]
```diff
@@ -117,6 +110,7 @@ mod tests {
 
         assert_eq!(ballot.description(&admin_ctx).unwrap(), description);
         assert_eq!(ballot.candidates(&admin_ctx).unwrap(), candidates);
+        // mantle_test::ext::mantle_err_len(std::ptr::null_mut());
```
Comment (armaniferrante) [inline examples/ballot/src/lib.rs]: Delete?

File: mantle-build/tests/ui/generic_state.rs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+#[derive(mantle::Service, Default)]
+pub struct State<T>(Option<T>);
+
+impl<T: Default> State<T> {
+    pub fn new(ctx: &Context) -> Result<Self, ()> {
+        Ok(Default::default())
+    }
+
+    fn hmmm() -> Result<(), String> {
```
Comment (armaniferrante) [inline mantle-build/tests/ui/generic_state.rs]: hm 🤔

File: blockchain-traits/src/lib.rs [hunk 6]
```diff
@@ -42,7 +42,7 @@ pub trait Blockchain: KVStore {
 
     /// Returns the input provided by the calling context.
     fn fetch_input(&self) -> Vec<u8>;
-    fn input_len(&self) -> u64;
+    fn input_len(&self) -> u32;
```
[the remark this answers is not in the document]
Comment (nhynes) [reply blockchain-traits/src/lib.rs]: Wasm is 32 bits. `usize` is `u32`. There is absolutely no way to have a 64-bit address space.

File: examples/ballot/src/lib.rs (hunk 4 above)
Comment (nhynes) [reply examples/ballot/src/lib.rs]: good catch. this was a feeble attempt to get rustc to link in `mantle_test`

File: examples/ballot/Cargo.toml [hunk 7]
```diff
@@ -9,14 +9,8 @@ publish = false
 crate-type = ["rlib", "cdylib"]
 
 [dependencies]
-mantle = { git = "https://github.com/oasislabs/mantle" }
+mantle = { git = "https://github.com/oasislabs/mantle", branch = "plugin" }
```
Comment (nhynes) [inline examples/ballot/Cargo.toml]:
```suggestion
mantle = { git = "https://github.com/oasislabs/mantle" }
```

File: examples/ballot/Cargo.toml [hunk 8]
```diff
@@ -9,14 +9,8 @@ publish = false
...
 
 [dev-dependencies]
-mantle-test = { git = "https://github.com/oasislabs/mantle" } # required for tests
-
-[build-dependencies]
-mantle = { git = "https://github.com/oasislabs/mantle" } # required for build script
-
-[features]
-deploy = [] # (automatically) set when building deployable service Wasm binary
+mantle-test = { git = "https://github.com/oasislabs/mantle", branch = "plugin" }
```
Comment (nhynes) [inline examples/ballot/Cargo.toml]:
```suggestion
mantle-test = { git = "https://github.com/oasislabs/mantle" }
```

File: examples/messaging/Cargo.toml [hunk 9]
```diff
@@ -9,14 +9,8 @@ publish = false
 crate-type = ["rlib", "cdylib"]
 
 [dependencies]
-mantle = { git = "https://github.com/oasislabs/mantle" }
+mantle = { git = "https://github.com/oasislabs/mantle", branch = "plugin" }
```
Comment (nhynes) [inline examples/messaging/Cargo.toml]:
```suggestion
mantle = { git = "https://github.com/oasislabs/mantle" }
```

File: examples/messaging/Cargo.toml [hunk 10]
```diff
@@ -9,14 +9,8 @@ publish = false
...
 
 [dev-dependencies]
-mantle-test = { git = "https://github.com/oasislabs/mantle" }
-
-[build-dependencies]
-mantle = { git = "https://github.com/oasislabs/mantle" }
-
-[features]
-deploy = []
+mantle-test = { git = "https://github.com/oasislabs/mantle", branch = "plugin" }
```
Comment (nhynes) [inline examples/messaging/Cargo.toml]:
```suggestion
mantle-test = { git = "https://github.com/oasislabs/mantle" }
```

**Pull request merged**